pengine/native.c

```diff
@@ -1342,30 +1342,41 @@ native_internal_constraints(resource_t * rsc, pe_working_set_t * data_set)
 
     if (is_stonith == FALSE
         && is_set(data_set->flags, pe_flag_enable_unfencing)
-        && is_set(rsc->flags, pe_rsc_needs_unfencing)
-        && is_not_set(rsc->flags, pe_rsc_have_unfencing)) {
+        && is_set(rsc->flags, pe_rsc_needs_unfencing)) {
         /* Check if the node needs to be unfenced first */
         node_t *node = NULL;
         GHashTableIter iter;
 
-        if(rsc != top) {
-            /* Only create these constraints once, rsc is almost certainly cloned */
-            set_bit_recursive(top, pe_rsc_have_unfencing);
-        }
-
         g_hash_table_iter_init(&iter, rsc->allowed_nodes);
         while (g_hash_table_iter_next(&iter, NULL, (void **)&node)) {
             action_t *unfence = pe_fence_op(node, "on", TRUE, data_set);
 
-            custom_action_order(top, generate_op_key(top->id, top == rsc?RSC_STOP:RSC_STOPPED, 0), NULL,
-                                NULL, strdup(unfence->uuid), unfence,
-                                pe_order_optional, data_set);
+            crm_debug("Ordering any stops of %s before %s, and any starts after",
+                      rsc->id, unfence->uuid);
 
-            crm_debug("Stopping %s prior to unfencing %s", top->id, unfence->uuid);
+            /*
+             * It would be more efficient to order clone resources once,
+             * rather than order each instance, but ordering the instance
+             * allows us to avoid unnecessary dependencies that might conflict
+             * with user constraints.
+             *
+             * @TODO: This constraint can still produce a transition loop if the
+             * resource has a stop scheduled on the node being unfenced, and
+             * there is a user ordering constraint to start some other resource
+             * (which will be ordered after the unfence) before stopping this
+             * resource. An example is "start some slow-starting cloned service
+             * before stopping an associated virtual IP that may be moving to
+             * it":
+             *       stop this -> unfencing -> start that -> stop this
```

Comment:
We've got three actions: start A, stop B, and unfence N1.

The cluster orders stops before unfencing, and starts after unfencing, so we have stop B -> unfence N1 -> start A. The user has a constraint start-A-then-stop-B, so it loops.

It's an unlikely scenario, because the only time there should be a stop and an unfencing on the same node is when the fence device parameters were just changed to include unfencing. So the problem should only happen if the cluster happens to be stopping the constrained resource at the same time.

The asymmetric "start-A-then-stop-B" is an unusual design, but legitimately useful. In the BZ case, A is a slow-starting clone, and B is a virtual IP colocated with it. If there is a node preference, there is an outage when the node comes online, because the IP stops immediately and doesn't start until A is finished starting. The constraint avoids that.

So ... we could narrow the problem space even further, by supporting a "move" ordering action, i.e. start-A-then-move-B. Move == if a resource is stopping on one node and starting on another, constrain both the stop and start. Or equivalently, we could automatically interpret start-then-stop constraints in this fashion. But that might be more confusing than helpful, and theoretically there could be other use cases for start-A-then-stop-B.

I think this is a case of the general problem that it's possible for a cluster-generated ordering to conflict with a user-configured ordering. The opposite of a cluster-generated ordering rarely makes sense, so we don't see it often.